the-collab-lab/tcl-66-smart-shopping-list
# 18. Navigation Bar

## Description
I worked in 4 files to build the navigation bar:
1. src/components/NavigationBar.jsx
2. src/components/NavigationBarAddList.jsx
3. src/components/NavigationBarModal.jsx
4. src/components/NavigationBarSingleList.jsx

- The navigation bar contains a New List button in the upper right corner that, when clicked, displays a form for the user to create a new list. 
- The navigation bar contains a sign out button in the lower-left corner.
- The navigation bar contains a “My Lists” section and a “[Lists] Shared With Me” section. If there are no shopping lists in the user’s database (either created by them or shared with them), “No Lists” is displayed in both sections.

**My Lists**
The signed-in user’s lists display in the My Lists section when the user’s uid matches the uid contained in the listPath for a shopping list. Two methods (find and map) make this work. The find method iterates through the user’s lists in the firestore database and one of two things will occur:

1. If the user’s uid does not match any of the uids contained in all listPaths, then “No Lists” is displayed.
2. If the user’s uid matches at least one uid in the listPaths for the shopping lists, then a map method will find all shopping lists that have a matching uid in their listPath. The user’s lists are displayed based on the map method’s execution.

**Shared With Me**
The signed-in user’s “Shared With Me” lists display in the respective section when the user’s uid differs from the uid contained in the listPath for a shopping list. A find method and map method make this work. The find method iterates through the user’s lists in the firestore database and one of two things will happen:
1. If the user’s uid matches all of the uids contained in all listPaths, then “No Lists” is displayed.
2. If the user’s uid differs from at least one uid in the listPaths for the shopping lists, then a map method will find all shopping lists that have a differing uid in their listPath. The shared lists are displayed based on the map method’s execution.

## Related Issue
Closes #18

## Acceptance Criteria
- [X] List component
- [X] Shared list component

## Updates
### Before
![before-nav-bar](https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/105468537/28755ebf-545f-44d4-98b3-d8008a7302ee)

### After
![Screen Shot 2024-04-01 at 12 48 25 PM](https://github.com/the-collab-lab/tcl-66-smart-shopping-list/assets/105468537/9f5217e8-0c8c-46ed-b7ef-04c9fb861d30)

## Testing Steps / QA Criteria

1. git pull
2. git checkout ec-issue-18
3. npm start

Lists will be displayed in the two sections in the navigation bar according to the above criteria. See the following steps for testing these sections, the New List button, and the Sign Out button.

**Testing "My Lists" Section**
1. Open [firestore console ](https://console.firebase.google.com/u/0/project/tcl-66-smart-shopping-list/firestore/data/~2Fusers~2Fdaheber@gmail.com) and navigate to your "users" document (named with your email address).
2. If you have lists in your collection (under the sharedLists array), confirm that they are displaying in the navigation bar in the "My Lists" section. If there are no lists in your collection, check the app to see that "No Lists" is displayed. 
3. Add at least one list to your collection by using "New List" button in the navigation bar of the application. Confirm that it appears in the "My Lists" section.
4. Click on a list inside of the "My Lists" section to be directed to that list's web page.

**Testing "Shared With Me" Lists Section**
1. Open [firestore console ](https://console.firebase.google.com/u/0/project/tcl-66-smart-shopping-list/firestore/data/~2Fusers~2Fdaheber@gmail.com) and navigate to your "users" document (named with your email address). If there are no lists shared with you by other users as indicated in the sharedLists array in your collection, check the app to see that "No Lists" is displayed and then go to step 3. If lists have been shared with you, skip to step 4.
3. In your sharedLists array, add a reference to a shopping list's path that is owned by another user. For example, if I wanted to shared Aloe's "Tools" list with myself, I would create this reference in my sharedLists array: [Aloe's uid]/Tools. The format of the reference is: [list owner's uid]/[List Name]. 
4. Lists that are shared with you should appear in your "Shared With Me" section in the app. 
5. Click on a list in the navigation bar to be directed to the respective list page.
7. In your collection inside the firestore console, delete all lists that have been shared with you by deleting them from your sharedLists array. Then check the application to confirm that "No Lists" is displayed.

**Testing "Sign Out" Button**
1. Click on the Sign Out button in the lower-left corner of the navigation bar to be signed out of the application.
